### PR TITLE
62 neon cli emulate external call

### PR DESCRIPTION
**Pull request by vasiliy-zaznobin**
If an external call solana.call() is made in a contract, then emulating a call to such a contract led to an error.
Such a call may not always be possible outside of the BPF virtual machine.
Therefore, such a call is implemented in evm_loader::solana_backend pub trait AccountStorage as trait, which has two implementations:
one for the BPF virtual machine evm_loader::account_storage pub struct ProgramAccountStorage<'a>,
and the other for emulation neon_cli::account_storage pub struct EmulatorAccountStorage<'a>.

### CODE REVIEW

**Review by otselnik — APPROVED** [review]

File: evm_loader/ERC20/test/test_erc20.py [hunk 1]
```diff
@@ -310,6 +310,21 @@ def __send_neon_transaction(self, evm_trx_data, ethereum_transaction, need_stora
...
+    print('sender:', sender)
+    print('contract:', contract)
+    print('trx_data:', trx_data)
+    cli = neon_cli()
+    cli_result = cli.emulate(evm_loader_id, sender + ' ' + contract + ' ' + trx_data)
+    print('cli_result:', cli_result)
+    emulate_result = json.loads(cli_result)
+    print('emulate_result:', emulate_result)
+    assert (emulate_result["exit_status"] == 'succeed')
+    assert(int(emulate_result['result']) == 1)
```
Comment (s-medvedev) [inline evm_loader/ERC20/test/test_erc20.py:324]: Please add checks for returned accounts info from the emulator.

File: evm_loader/ERC20/test/test_erc20.py [hunk 2]
```diff
@@ -421,8 +436,14 @@ def test_erc20(self):
         assert (erc20.balance(self.ethereum_caller) == 0)
 
         deposit_amount = 1
-        erc20.deposit(self.ethereum_caller, client_acc, self.ethereum_caller, deposit_amount * (10 ** 9),
-                      balance_erc20, token, self.acc.public_key()._key)
+        deposit_trx_data = erc20.deposit(self.ethereum_caller, client_acc, self.ethereum_caller,
+                                         deposit_amount * (10 ** 9), balance_erc20, token,
+                                         self.acc.public_key()._key)
+
+        check_deposit_emulation(self.ethereum_caller.hex(),
```
Comment (s-medvedev) [inline evm_loader/ERC20/test/test_erc20.py:443]: I suggest writing separate tests for check emulation. ERC20 main functional and emulator refer to different functionality. Ideally, tests should show what is not working.

File: evm_loader/ERC20/test/test_erc20.py (hunk 1 above)
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test_erc20.py]: Fixed. Add checks of the accounts info

File: evm_loader/ERC20/test/test_erc20.py (hunk 2 above)
Comment (vasiliy-zaznobin) [reply evm_loader/ERC20/test/test_erc20.py]: Fix. Introduce test_cli_emulate.py

File: evm_loader/test_cli_emulate.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        checking_account = [item["account"], item["address"], item["contract"]]
+        exists = checking_account in accounts_should_be
+        print('checking_account:', checking_account, exists)
+        accounts_not_found -= exists
+    print('accounts_not_found:', accounts_not_found)
+    assert (accounts_not_found == 0)
+    print('deposit emulation: OK\n')
+
+
+class ERC20test(unittest.TestCase):
```
Comment (s-medvedev) [inline evm_loader/test_cli_emulate.py:81]: Fix ERC20test name

File: evm_loader/test_cli_emulate.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        self.tokenMint(token, token_acc1, mint_amount)
+        assert (self.tokenBalance(token_acc1) == mint_amount)
+        assert (self.tokenBalance(token_acc2) == 0)
+
+        transfer_amount = 17
+        trx_data = self.contract.transfer_ext(self.ethereum_caller,
+                                              token, token_acc1, token_acc2, transfer_amount * (10 ** 9),
+                                              self.acc.public_key()._key)
+
+        assert (self.tokenBalance(token_acc1) == mint_amount - transfer_amount)
```
Comment (s-medvedev) [inline evm_loader/test_cli_emulate.py:180]: Please, use assertions from unittest.

File: evm_loader/test_cli_emulate.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    cli = neon_cli()
+    cli_result = cli.emulate(evm_loader_id, sender + ' ' + contract + ' ' + trx_data)
+    print('cli_result:', cli_result)
+    emulate_result = json.loads(cli_result)
+    print('emulate_result:', emulate_result)
+    assert (emulate_result["exit_status"] == 'succeed')
+    assert (emulate_result['result'] == '')  # no type return from transferExt
+    print('accounts_should_be:', accounts_should_be)
+    accounts_not_found = len(accounts_should_be)
+    for item in emulate_result["accounts"]:
```
Comment (s-medvedev) [inline evm_loader/test_cli_emulate.py:71]: Check all accounts.

File: evm_loader/solana_utils.py [hunk 6]
```diff
@@ -35,6 +43,151 @@
...
+        print('cmd:', cmd)
+        try:
+            return subprocess.check_output(cmd, shell=True, universal_newlines=True)
+        except subprocess.CalledProcessError as err:
+            import sys
+            print("ERR: spl-token error {}".format(err))
+            raise
+
+
+class EthereumTransaction:
```
[the remark this answers is not in the document]
Comment (vasiliy-zaznobin) [reply evm_loader/solana_utils.py:61]: Fixed

File: evm_loader/test_cli_emulate.py (hunk 4 above)
Comment (vasiliy-zaznobin) [reply evm_loader/test_cli_emulate.py]: Fixed

**Review by vakond — DISMISSED** [review]

**Review by vakond — APPROVED** [review]

File: evm_loader/test_cli_emulate.py (hunk 3 above)
Comment (vasiliy-zaznobin) [reply evm_loader/test_cli_emulate.py]: Fix

**Review by vakond — APPROVED** [review]

File: evm_loader/test_cli_emulate.py [hunk 7]
```diff
@@ -106,94 +97,138 @@ def setUpClass(cls):
...
+            ],
+            "exit_status": "succeed",
+            "result": ""
+        }"""
+        tmpl = tmpl.replace('CONTRACT_ACCOUNT', self.contract.contract_account)
+        tmpl = tmpl.replace('CONTRACT_ETHEREUM_ID', '0x' + self.contract.ethereum_id.hex())
+        tmpl = tmpl.replace('CONTRACT_CODE_ACCOUNT', self.contract.contract_code_account)
+        tmpl = tmpl.replace('ETHEREUM_CALLER_ACCOUNT', self.caller)
+        tmpl = tmpl.replace('ETHEREUM_CALLER_ID', '0x' + self.ethereum_caller.hex())
+        tmpl_json = json.loads(tmpl)
```
Comment (s-medvedev) [inline evm_loader/test_cli_emulate.py:172]: Why do you need to prepare `tmpl_json` from a string? You can build `tmpl_json` object in place like this:
```
        tmpl = {
            "accounts": [
                {
                    "account": self.contract.contract_account,
                    "address": '0x' + self.contract.ethereum_id.hex(),
                    "code_size": None,
                    "contract": self.contract.contract_code_account
                    "new": False,
                    "writable": True
                },
                {
                    "account": self.caller,
                    "address": '0x' + self.ethereum_caller.hex(),
                    "code_size": None,
                    "contract": None,
                    "new": False,
                    "writable": True
                }
            ],
            "exit_status": "succeed",
            "result": ""
        }
```

**Review by s-medvedev — APPROVED** [review]

File: evm_loader/test_cli_emulate.py (hunk 7 above)
Comment (vasiliy-zaznobin) [reply evm_loader/test_cli_emulate.py:172]: it will be resolved in a next task

**Pull request merged**